identity-core/src/crypto/exchange/x25519.rs

```diff
@@ -0,0 +1,59 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use core::marker::PhantomData;
+
+use crypto::keys::x25519;
+
+use crate::crypto::KeyExchange;
+use crate::Error;
+
+/// An implementation of `X25519` Elliptic-curve Diffie-Hellman (ECDH) cryptographic key exchange.
+// TODO: this phantom data workaround sucks, maybe make key_exchange generic instead?
```

Comment:
Thanks for the suggestions!

> Do we know sufficient other key exchange types to design an abstraction for this now? 

The only other ones I'm aware of are RSA, which has broadly been phased out in favour of elliptic-curves which are faster with smaller key sizes, so I don't think we'll have to support it. 

The problem with using generics directly rather than associated types is that generics mostly prevent specifying a particular struct, which associated types do.

E.g. `PUB: AsRef<X25519Public>` only works is `AsRef` is implemented manually for `X25519Public` since it is not a default trait, while `PUB: X25519Public` throws an error. An associated type such as `type PUB = X25519Public` works fine, however. I believe this is why existing traits like `Sign`, for instance, use associated types for input parameters.

There are three option basically:

1. Keep the `KeyExchange` trait with associated types and `PhantomData` in `X25519`.
2. Change the `KeyExchange` trait to use generics for input parameters, removing the need for `PhantomData`.
3. Remove the `KeyExchange` trait entirely and implement `key_exchange` for `X25519` directly.

I'm still very undecided, any of the options would be acceptable in my opinion.

Reply:
> I believe this is why existing traits like `Sign`, for instance, use associated types for input parameters.

Yeah, that make sense.

I'm in favor of option 3. Implementing an interface that we may have to break in the future, because its abstraction was based on just this case, is probably not worth that much. I'd go with the simple option for now and forgo the trait.
If you want to keep the trait, option 2 sounds somewhat preferable to option 1.

Reply:
Removed the `KeyExchange` trait, moved `key_exchange` to `X25519`.